### PR TITLE
[CPU] Add RDFT and IRDFT operators

### PR DESCRIPTION
**Pull request by mateusztabaka**
Tickets: 79178 and 79192

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 1]
```diff
@@ -0,0 +1,1537 @@
...
+static const size_t SIGNAL_SIZE_INDEX = 2;
+
+
+bool RDFT::isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op, std::string& errorMessage) noexcept {
+    try {
+        if (isDynamicNgraphNode(op)) {
+            errorMessage = "Doesn't support op with dynamic shapes";
+            return false;
+        }
+        const bool is_rdft = is_type<const ov::op::v9::RDFT>(op);
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:47]: Please use camelCase for variables and functions.

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 2]
```diff
@@ -0,0 +1,1537 @@
...
+    return sizeof(T) * 2;
+}
+
+enum dft_type {
+    real_to_complex,
+    complex_to_complex,
+    complex_to_real,
+};
+
+#define GET_OFF(field) offsetof(jit_dft_args, field)
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:185]: Please move all JIT related code to separate file in folder intel_cpu/src/nodes/kernels

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 3]
```diff
@@ -0,0 +1,1537 @@
...
+            errorMessage = "Only opset9 RDFT/IRDFT operation is supported";
+            return false;
+        }
+    } catch (...) {
+        return false;
+    }
+    return true;
+}
+
+static void normalize_axes(std::vector<int>& axes, size_t rank, bool inverse) {
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:60]: Signal size input must be taken into account here as well if it exists.

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 4]
```diff
@@ -0,0 +1,1537 @@
...
+            rdft_kernel.reset(new jit_dft_kernel_f32<cpu::x64::avx2>(is_inverse, rdft_type));
+            dft_kernel.reset(new jit_dft_kernel_f32<cpu::x64::avx2>(is_inverse, complex_to_complex));
+            fft_kernel.reset(new jit_fft_kernel_f32<cpu::x64::avx2>(is_inverse));
+            vlen = cpu_isa_traits<cpu::x64::avx2>::vlen;
+        } else if (mayiuse(cpu::x64::sse41)) {
+            rdft_kernel.reset(new jit_dft_kernel_f32<cpu::x64::sse41>(is_inverse, rdft_type));
+            dft_kernel.reset(new jit_dft_kernel_f32<cpu::x64::sse41>(is_inverse, complex_to_complex));
+            vlen = cpu_isa_traits<cpu::x64::sse41>::vlen;
+        } else {
+            IE_THROW() << "Can't create RDFT kernel";
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:1125]: There can be two solutions:
1. Implement some reference here in the CPU plug-in,
2. Add the same check in the function _isSupportedOperation_ and fallback on core reference.

**Comment by nshchego** [comment]
@mateusztabaka, are you going to add CPU specific tests here? They are required to confirm sustainability on different platforms.

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 5]
```diff
@@ -0,0 +1,1537 @@
...
+
+    std::vector<PortConfigurator> configurators({{LayoutType::ncsp, Precision::FP32},
+                                                 {LayoutType::ncsp, Precision::I32}});
+    if (inputShapes.size() > SIGNAL_SIZE_INDEX)
+        configurators.push_back({LayoutType::ncsp, Precision::I32});
+
+    addSupportedPrimDesc(configurators, {{LayoutType::ncsp, Precision::FP32}}, impl_desc_type::ref_any);
+}
+
+void RDFT::execute(dnnl::stream strm) {
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:132]: Do you use some known algorithm for RDFT here? As I understand, imaginary part is filled by zero and Cooley Tukey FFT approach is implemented here?

File: src/bindings/python/tests/test_ngraph/test_rdft.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+def test_rdft_1d():
+    runtime = get_runtime()
+    n = 50
+    shape = [n]
+    data = np.random.uniform(0, 1, shape).astype(np.float32)
+    param = ov.parameter(Shape(shape), name="input", dtype=np.float32)
+    input_axes = ov.constant(np.array([0], dtype=np.int64))
+
+    node = ov.rdft(param, input_axes)
```
Comment (jiwaszki) [inline src/bindings/python/tests/test_ngraph/test_rdft.py:20]: I see tests for 2D-rdft with `signal_size` param, is it by design that 1D case is missing?

**Review by jiwaszki — COMMENTED** [review]
Overall Python tests are legit, just one small question about tests.

File: src/bindings/python/tests/test_ngraph/test_rdft.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    signal_size = [33, 64]
+    data = np.random.uniform(0, 1, shape).astype(np.float32)
+    param = ov.parameter(Shape(shape), name="input", dtype=np.float32)
+    axes = [-2, -1]
+    input_axes = ov.constant(np.array(axes, dtype=np.int64))
+    signal_size_node = ov.constant(np.array(signal_size, dtype=np.int64))
+    node = ov.irdft(param, input_axes, signal_size_node)
+    computation = runtime.computation(node, param)
+    actual = computation(data)
+    np_results = np.fft.rfftn(data, axes=axes)
```
Comment (jiwaszki) [inline src/bindings/python/tests/test_ngraph/test_rdft.py:131]: One more thing, this `np_results` are not used, should they be used in expected results as `data`?

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 8]
```diff
@@ -0,0 +1,1753 @@
...
+            size_t block_size = 4 * cacheline_size / sizeof(float);
+            size_t num_blocks = (output_size + block_size - 1) / block_size;
+            parallel_nt(num_blocks, [&] (size_t i, size_t nthr) {
+                if (num_blocks > nthr) {
+                    auto new_block_size = (((output_size / nthr) + block_size - 1) / block_size) * block_size;
+                    block_size = new_block_size;
+                    num_blocks = nthr;
+                }
+                jit_dft_args args {
+                    .input = input_ptr,
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:1393]: We support just C++11 compiler version now. Looks like such syntax is not supported by this compiler.

File: src/plugins/intel_cpu/src/nodes/rdft.cpp (hunk 3 above)
Comment (mateusztabaka) [reply src/plugins/intel_cpu/src/nodes/rdft.cpp]: actually the sorting was unnecessary

File: src/plugins/intel_cpu/src/nodes/rdft.cpp (hunk 5 above)
Comment (mateusztabaka) [reply src/plugins/intel_cpu/src/nodes/rdft.cpp:123]: > As I understand, imaginary part is filled by zero and Cooley Tukey FFT approach is implemented here?
right, FFT is based on Cooley Tukey and the imag part is filled with zeros. For non power-of-two cases, naive DFT is implemented and it works directly on real data.

File: src/bindings/python/tests/test_ngraph/test_rdft.py (hunk 6 above)
Comment (mateusztabaka) [reply src/bindings/python/tests/test_ngraph/test_rdft.py:20]: It's by design. Signal size test cases are just to make sure python bindings respect that parameter in the API.

File: src/bindings/python/tests/test_ngraph/test_rdft.py (hunk 7 above)
Comment (mateusztabaka) [reply src/bindings/python/tests/test_ngraph/test_rdft.py]: right, this call is unnecessary

**Review by mbencer — APPROVED** [review]

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 9]
```diff
@@ -0,0 +1,939 @@
...
+                      twiddles, rank,
+                      axes, signalSizes,
+                      inputShape, outputShape,
+                      inputStrides, outputStrides);
+}
+
+bool RDFT::created() const {
+    return getType() == Type::RDFT;
+}
+
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:201]: Why did you override the createPrimitive and needPrepareParams functions?

File: src/plugins/intel_cpu/src/nodes/rdft.cpp [hunk 10]
```diff
@@ -0,0 +1,939 @@
...
+using namespace dnnl;
+using namespace InferenceEngine;
+
+namespace ov {
+namespace intel_cpu {
+namespace node {
+
+
+static const size_t DATA_INDEX = 0;
```
Comment (nshchego) [inline src/plugins/intel_cpu/src/nodes/rdft.cpp:32]: Define these constants as constexpr as well.

File: src/plugins/intel_cpu/src/nodes/rdft.cpp (hunk 9 above)
Comment (mateusztabaka) [reply src/plugins/intel_cpu/src/nodes/rdft.cpp:201]: right, it's redundant

**Pull request merged**